rust/src/client.rs

```diff
@@ -0,0 +1,865 @@
+            // Safety: Created via Box::into_raw
+            let _ = Box::from_raw(self.api);
+            settings_destroy(&mut self.ctx);
+        }
+    }
+}
+
+unsafe impl Send for ClientInner<'_> {}
+unsafe impl Sync for ClientInner<'_> {}
+
+impl<'a> Client<'a> {
+    pub fn new<F>(link: Link<'a, ()>, sender: F) -> Client<'a>
+    where
+        F: FnMut(SBP) -> Result<(), Box<dyn std::error::Error + Send + Sync>> + 'static,
+    {
+        let context = Box::new(Context {
+            link,
+            sender: Box::new(sender),
+            callbacks: Vec::new(),
+            lock: Lock::new(),
+            event: None,
+        });
+
+        let api = Box::new(settings_api_t {
+            ctx: ptr::null_mut(),
+            send: Some(libsettings_send),
+            send_from: Some(libsettings_send_from),
+            wait_init: Some(libsettings_wait_init),
+            wait: Some(libsettings_wait),
+            wait_deinit: None,
+            signal: Some(libsettings_signal),
+            wait_thd: Some(libsettings_wait_thd),
+            signal_thd: Some(libsettings_signal_thd),
+            lock: Some(libsettings_lock),
+            unlock: Some(libsettings_unlock),
+            register_cb: Some(register_cb),
+            unregister_cb: Some(unregister_cb),
+            log: None,
+            log_preformat: false,
+        });
+
+        let mut inner = Box::new(ClientInner {
+            context: ptr::null_mut(),
+            api: Box::into_raw(api),
+            ctx: ptr::null_mut(),
+        });
+
+        let context_raw = Box::into_raw(context);
+        inner.context = context_raw;
+
+        // Safety: inner.api was just created via Box::into_raw so it is
+        // properly aligned and non-null
+        unsafe {
+            (*inner.api).ctx = context_raw as *mut c_void;
+        }
+
+        inner.ctx = unsafe { settings_create(SENDER_ID, inner.api) };
+        assert!(!inner.ctx.is_null());
+
+        Client { inner }
+    }
+
+    pub fn read_all(&self) -> Vec<Result<ReadByIdxResult, Error<ReadSettingError>>> {
+        const NUM_WORKERS: usize = 5;
+
+        thread::scope(move |scope| {
+            let (idx_s, idx_r) = crossbeam_channel::bounded(NUM_WORKERS);
+            let (results_s, results_r) = crossbeam_channel::unbounded();
+
+            scope.spawn(move |_| {
+                let mut idx = 0;
+                while idx_s.send(idx).is_ok() {
+                    idx += 1;
+                }
+            });
+
+            for _ in 0..NUM_WORKERS {
+                let idx_r = idx_r.clone();
+                let results_s = results_s.clone();
+                scope.spawn(move |_| {
+                    for idx in idx_r.iter() {
+                        match self.read_by_index(idx).map_err(Error::Err).transpose() {
+                            Some(res) => results_s.send(res).expect("results channel closed"),
+                            None => break,
+                        }
```

Comment:
Can we confirm that this is multi-threaded?  Like add an `AtomicU32` in the outer scope (or globally) then increment while `read_by_index` is running?

E.g.
```suggestion
                        COUNTER.fetch_add(1, Ordering::SeqCst);
                        println!("threads running: {}", some_var.load(Ordering::SeqCst));
                        match self.read_by_index(idx).map_err(Error::Err).transpose() {
                            Some(res) => results_s.send(res).expect("results channel closed"),
                            None => break,
                        }
                        COUNTER.fetch_sub(1, Ordering::SeqCst);
```

Reply:
Yeah just tried and seems so 

```
threads running: 1
threads running: 3
threads running: 2
threads running: 4
threads running: 5
....
```

Reply:
Ok, cool